src/asm/x86/quantize.rs

```diff
@@ -0,0 +1,214 @@
+        unsafe {
+          (func)(
+            qindex,
+            coeffs.as_ptr() as *const _,
+            eob,
+            rcoeffs.as_mut_ptr() as *mut _,
+            tx_size,
+            bit_depth,
+            dc_delta_q,
+            ac_delta_q,
+          )
+        }
+      } else {
+        call_native(rcoeffs)
+      }
+    }
+    PixelType::U16 => call_native(rcoeffs),
+  }
+
+  #[cfg(any(feature = "check_asm", test))]
+  {
+    let area = av1_get_coded_tx_size(tx_size).area();
+    assert_eq!(&rcoeffs[..area], &ref_rcoeffs[..]);
+  }
+}
+
+#[target_feature(enable = "avx2")]
+unsafe fn dequantize_avx2(
+  qindex: u8, coeffs_ptr: *const i16, _eob: usize, rcoeffs_ptr: *mut i16,
+  tx_size: TxSize, bit_depth: usize, dc_delta_q: i8, ac_delta_q: i8,
+) {
+  let log_tx_scale = _mm256_set1_epi32(get_log_tx_scale(tx_size) as i32);
+
+  let quants_ac =
+    _mm256_set1_epi32(ac_q(qindex, ac_delta_q, bit_depth) as i32);
+  // Use the dc quantize as first vector element for the first iteration
+  let mut quants = _mm256_insert_epi32(
+    quants_ac,
+    dc_q(qindex, dc_delta_q, bit_depth) as i32,
+    0,
+  );
+
+  let area: usize = av1_get_coded_tx_size(tx_size).area();
+  // Step by 16 (256/16) coefficients for each iteration
+  let step: usize = 16;
+  assert!(area >= step);
+
+  // Increase the pointers as we iterate
+  let mut coeffs_ptr: *const i16 = coeffs_ptr;
+  let mut rcoeffs_ptr: *mut i16 = rcoeffs_ptr;
+  for _i in (0..area).step_by(step) {
+    let coeffs = _mm256_load_si256(coeffs_ptr as *const _);
+    let coeffs_abs = _mm256_abs_epi16(coeffs);
+
+    // TODO: Since log_tx_scale is at most 2 and we gain an extra bit by taking
+    // the abs value (unless the range is [-(2^15), 2^15 + 1]), it might be
+    // possible to perform a 16-bit multiply and get the highest bit by
+    // comparing coeffs to (1<<16) / quant. The would cost 1 compare, 1 blend,
+    // and 1 add, but would get rid of 1 pack, 2 unpacks, 1 shift, and 1
+    // multiply.
+
+    let rcoeffs = _mm256_sign_epi16(
+      _mm256_packs_epi32(
+        // (abs_coeff * quant) >> log_tx_scale
+        _mm256_srlv_epi32(
+          _mm256_madd_epi16(
+            quants,
+            // Convert the first half of each lane to 32-bits
+            _mm256_unpacklo_epi16(coeffs_abs, _mm256_setzero_si256()),
+          ),
+          log_tx_scale,
+        ),
+        // Second half
+        _mm256_srlv_epi32(
+          _mm256_madd_epi16(
+            quants_ac,
+            _mm256_unpackhi_epi16(coeffs_abs, _mm256_setzero_si256()),
+          ),
+          log_tx_scale,
+        ),
+      ),
+      coeffs,
+    );
+    _mm256_store_si256(rcoeffs_ptr as *mut _, rcoeffs);
+
+    // Only use a dc quantizer for the first iteration
+    quants = quants_ac;
+    coeffs_ptr = coeffs_ptr.add(step);
+    rcoeffs_ptr = rcoeffs_ptr.add(step);
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use super::*;
+  use rand::{thread_rng, Rng};
+
+  #[test]
+  fn dequantize_test() {
```

Comment:
This should have an effect only when check_asm is enabled, right?

Reply:
No, I enabled to results checking code to be enabled either for check_asm or test.

Reply:
Then I guess it is good to go :)